[2~4단계 방탈출 결제 / 배포] 엠제이(최민준) 미션 제출합니다

theminjunchoi opened the pull request:
안녕하세요 오리!
이번 미션도 리뷰 잘 부탁드립니다!

배포, 로깅, 문서화 모두 처음 학습해보았는데요, 자유롭게 피드백 부탁드립니다~!
지난 미션 마지막에 해주신 질문에 대해서는 코드에 답변 달아놓겠습니다~

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4415)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
```
[x] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
[x] else 예약어를 쓰지 않는다.
[x] 모든 원시 값과 문자열을 포장한다.
[ ] 일급 컬렉션을 쓴다.
[x] 한 줄에 점을 하나만 찍는다.
[x] 줄여 쓰지 않는다(축약 금지).
[ ] 모든 엔티티를 작게 유지한다.
[ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
[ ] getter/setter/프로퍼티를 쓰지 않는다.
```

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
### 배포, 문서화 관련
| 항목   | 링크 |
|--------|------|
| ERD    | [README](https://github.com/theminjunchoi/spring-roomescape-payment/blob/step2/README.md) |
| 배포   | [배포 사이트](http://3.36.117.160:8080/) |
| 문서화 | [Swagger API 문서](http://3.36.117.160:8080/swagger-ui/index.html) |

### ETC
코드에 질문 남겨놓겠습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

theminjunchoi committed:
docs: README에 결제 정보 표시 기능 추가 및 업데이트

theminjunchoi committed:
feat: 예약 내역에 결제 정보 표시 기능 추가

- Reservation 관련 응답 DTO(MyHistoryResponse)에 결제 관련 필드(paymentKey, amount) 추가
- PaymentRepository에 Reservation 기반 결제 조회 메서드(findByReservation) 추가
- Reservation 조회 시 결제 정보를 포함하도록 ReservationQueryService 로직 수정
- 테스트 데이터 및 테스트 케이스에서 결제 정보 추가
- 예약 페이지(reservation-mine.js, reservation-mine.html)에서 결제 정보 표시 기능 구현
- 관련 SQL 스크립트 수정 및 결제 더미 데이터 추가
- Payment 엔티티에서 paymentStatus로 필드명 변경 및 선언 수정

theminjunchoi committed:
refactor: 본인 예약 조회 시 paymentKey 안보여주도록 수정

theminjunchoi committed:
refactor: 결제 대기 로직에서 불필요한 데이터 제거 및 승인 처리 개선

- PaymentDataRequest를 제거하고 Reservation에서 결제 금액(amount) 조회하도록 변경
- TossPaymentService 및 관련 테스트 케이스 수정
- 결제 대기 상태를 MyHistoryResponse에 반영 및 상태에 따라 버튼 추가 (결제, 취소 등)
- ReservationCommandService에서 결제 승인 로직 간소화
- PaymentRepository에 다중 상태 기반 결제 조회 메서드 추가
- 관련 JS 코드 수정으로 프론트엔드 상태 처리 및 UI 개선

theminjunchoi committed:
feat: Swagger API 문서 기능 추가 및 API Operation 주석 명시

theminjunchoi committed:
docs: ERD 다이어그램 추가

theminjunchoi committed:
docs: ERD 다이어그램 수정

theminjunchoi committed:
docs: ERD 다이어그램 수정

theminjunchoi committed:
docs: ERD 다이어그램 수정

theminjunchoi committed:
docs: ERD 다이어그램 수정

theminjunchoi committed:
docs: ERD 다이어그램 수정

theminjunchoi committed:
refactor: 예외 처리 로직 개선

theminjunchoi committed:
refactor: 결제 재시도 로직 분리 및 관련 테스트 코드 수정

theminjunchoi committed:
refactor: 대기 승격 테스트명 변경

theminjunchoi committed:
refactor: 예외 처리 시 로그 레벨 및 메시지 포맷 수정

theminjunchoi committed:
refactor: 테스트 종료 후 MockWebServer 종료 로직 제거

theminjunchoi committed:
refactor: 사용안하는 코드 제거

theminjunchoi committed:
refactor: TossPaymentClient 예외 처리 로직 및 사용하지 않는 코드 제거

theminjunchoi committed:
refactor: TossPaymentClient 예외 처리 로직 개선 및 HttpComponentsClientHttpRequestFactory 적용

- 4xx 및 5xx 에러 처리 메서드 분리 및 추가 예외 처리 로직 적용
- HttpComponentsClientHttpRequestFactory로 요청 팩토리 변경
- 의존성에 Apache HttpClient 추가

theminjunchoi committed:
refactor: TossPaymentClient 예외 타입 변경

theminjunchoi commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -30,20 +30,28 @@ public PaymentResponse requestPayment(final PaymentRequest request) {
-            TossErrorResponse error = objectMapper.readValue(is, TossErrorResponse.class);
-            throw new TossPaymentException(HttpStatus.valueOf(error.code()), error.message());
+    private void handle4xxError(HttpRequest httpRequest, ClientHttpResponse clientHttpResponse) {
+        try {
+            JsonNode node = objectMapper.readTree(clientHttpResponse.getBody());
+            String code = node.path("code").asText();
+            String message = node.path("message").asText();
+            if (code.equals("UNAUTHORIZED_KEY")) {
+                throw new TossPaymentException(HttpStatus.INTERNAL_SERVER_ERROR,
+                        "방탈출 예약 서비스 결제 시스템에서 오류가 발생했습니다. 잠시 후 다시 시도해주세요.");
+            }
+            throw new TossPaymentException(HttpStatus.BAD_REQUEST, message);
         } catch (IOException e) {
-            throw new RuntimeException("에러 응답 파싱 실패", e);
+            throw new RuntimeException(e);
         }
     }
```
https://github.com/woowacourse/spring-roomescape-payment/pull/206#discussion_r2121051139

이 질문에 대해서도 고민을 해봤습니다.

실제로 secret key에 적합하지 않은 값을 넣어서 테스트를 해보았고, 이때 토스 api는 401, UNAUTHORIZED_KEY를 반환했습니다. 이를 제외한 4xx 에러는 모두 사용자의 잘못된 입력값에 대한 에러였습니다. 

그래서 secret key와 관련된 에러는 토스가 4xx 에러를 던져주더라도 사용자에게는 5xx 에러를 던져주도록 코드를 작성했습니다.

theminjunchoi commented on src/main/java/roomescape/payment/infrastructure/TossPaymentService.java:
```diff
@@ -48,18 +44,29 @@ public Payment pay(
+    )
+    public void paymentClientWithRetry(PaymentRequest request) {
+        paymentClient.requestPayment(request);
+    }
```
https://github.com/woowacourse/spring-roomescape-payment/pull/206#discussion_r2121079710

1번은 진짜 조심해야하는 부분이라 생각하고,
2번도 토스처럼 장애 시 스레드를 잡아먹는 구조라면 치명적이라고 생각합니다.

그래서 단순히 반복문을 돌면서 정해진 횟수안에서 시도해보는 것도 생각을 해봤고, 재시도가 필요한 부분과 아닌 부분으로 나누는 방법도 떠올랐는데, API 서버나 결제 서버가 간헐적으로 응답 느릴 때처럼 간헐적으로 네트워크 오류가 발생했을 때를 위해서 후자의 방식을 택해봤습니다.

지금도 계속 고민을 해보고 있지만, retryable이 꼭 필요한 때가 언제일까요?

jinyoungchoi95 replied:
지난 리뷰에서 말씀드렸지만 재시도 가능여부는 내가 관리할 수 있는 영역에 호출하느냐에 따라서 달라집니다. 외부 서버에 대해서라면 응답이 느리거나 간헐적으로 네트워크 오류가 발생할 것 같으니 임의로 판단해서 무작정 retry를 거는 것이 아니라 상대 서버가 retry를 해도 되는 것인지에 대한 상호 합의가 있어야해요.

현재처럼 토스와의 상호합의가 없는 상황이라면 retry를 지양합니다. 외부서버에 호출하는 것이라면 retry 사용 여부를 선택하는 것은 엠제이가 만든 서버가 아니라 호출받는 서버가 정하는 것이 맞는 것 같네요. 서로에게 장애 전파만 유도시킬뿐이에요. 

> retry는 일반적으로 "내가 제어할 수 있는 범위인가"와 "여러번 했을 때 부담이 덜한가"를 종합적으로 판단하여 설정해야합니다.

지난 리뷰에서 말씀드린 것처럼 서버가 본인이 제어할 수 없는 범위에 대해서는 고려하지 않습니다. 필요하다라고 특정한 상황을 정해놓지는 않는 것 같네요.

theminjunchoi replied:
아하 이해됐습니다! 감사합니다

theminjunchoi commented on src/main/java/roomescape/reservation/application/ReservationQueryService.java:
```diff
@@ -79,10 +82,20 @@ public List<MyHistoryResponse> findMyReservation(final Long memberId) {
 
         final List<Reservation> reservations = reservationRepository.findByMemberIdWithAssociations(memberId);
-        reservations.forEach(r -> responses.add(MyHistoryResponse.ofReservation(r)));
+        for (Reservation reservation : reservations) {
+            responses.add(
+                    MyHistoryResponse.ofReservation(reservation,
+                            paymentRepository.findByReservationAndPaymentStatusIn(
+                                    reservation,
+                                    List.of(PaymentStatus.AWAIT, PaymentStatus.SUCCESS)
+                            )
+                    )
+            );
```
코드래빗도 짚어주고 있는 문제이지만, 저는 우선 현재 코드도 괜찮다고 생각했어요.

현재 로직은 멤버의 id로 예약을 조회하고, 각 예약을 순회하면서 예약과 연관관계가 있는 결제 객체들을 찾고 있습니다.
물론 이 방법 말고, paymentRepository에 jpql이나 native query를 작성해서 한 번에 필요한 객체들을 찾아올 수도 있을 것 같은데, 
사용자 개인의 예약 기록을 찾기 위해 paymentRepository에게 묻는 게 어색하다고 느껴졌습니다.
또한 사용자 개인의 예약 기록이 몇 개 없을텐데, 이 쿼리 수를 줄이자고 객체지향적으로 어색해지는 것 같았습니다.

객체지향과 쿼리 수 개선이라는 문제를 동시에 달성하면 물론 좋겠지만, 선택을 해야한다면 기준을 어디에 둬야할까요?
저는 위에서도 말씀드린 것처럼 사용자 개인의 예약이 많지 않을 것 같아서 올바른 객체에게 묻도록 코드를 작성했습니다.

jinyoungchoi95 replied:
우선 객체지향과 쿼리에 대한 질문에서 먼저 고려해야하는 것에 대한 답은 쿼리입니다. 객체지향을 안지킨다고 장애를 발생시키지는 않지만 쿼리는 비정상적으로 호출하게될 경우 장애를 발생시킵니다. 지금까지는 성능을 별개로보고 객체지향을 지켰다면 앞으로는 서비스를 운영하는 입장에서 어떻게 장애가 나지 않는 서비스를 만들 것인가에도 초점을 맞추어보시면 좋을 것 같네요.

현재의 경우에는 그냥 `Reservation`를 조회할 때 `List<Payment>`를 같이 조회하게 하고, 거기서 값을 가져오면 되는 것 아닌가요? 말씀하신 객체지향의 관점을 말씀하셨는데 객체끼리 관계를 맺는 것이 아닌 연결을 끊어두셨는데 현재 코드는 객체지향을 지키고 있는 것이 맞을까요?

theminjunchoi replied:
> 지금까지는 성능을 별개로보고 객체지향을 지켰다면 앞으로는 서비스를 운영하는 입장에서 어떻게 장애가 나지 않는 서비스를 만들 것인가에도 초점을 맞추어보시면 좋을 것 같네요.

좋은 피드백 감사합니다.

> 현재의 경우에는 그냥 Reservation를 조회할 때 List<Payment>를 같이 조회하게 하고, 거기서 값을 가져오면 되는 것 아닌가요?

이 말씀도 맞는 것 같아요. 저도 이 생각을 안한 건 아니었어요.
하지만 여전히 고민은 되는 것 같습니다.

저희는 특정 유저의 예약 기록이 궁금한 상황이에요. 
이에 대해 답해줄 책임을 paymentRepository나 reservationRepository에게 부여해야하는데 솔직히 직관적으로는 reservationRepository에게 물어야할 것 같아요. 그리고 이런 생각을 기반으로 설계를 했을 때 추후 문제가 터져도 같은 생각으로 reservationRepository에 코드가 잘 짜여져 있는지 확인하러 올 것 같습니다. 

그리고 사용자 페이지에서는 해당 예약의 결제 상태에 대한 데이터도 필요한데, 이 부분은 따로 paymentRepository에게 물어보고, 저희는 이를 잘 조합해서 넘겨주면 된다고 생각했어요. 저는 이렇게 처리하는게 각 객체(XXXRepository)들에게 올바른 책임을 부여하는 것이라고 생각했습니다. 오히려 paymentRepository에게 List< Payment>를 받아오고 그 안에서 타고타고 들어가는 게 너무 많은 책임을 한 객체에게 부여하게 되는 꼴 아닌가? 하는 의심이 들었던 것 같습니다.

하지만 오리 말씀도 이해가 안가는 건 아니에요. 일단 문제 없이 서비스를 운영하는 게 먼저라고 생각하고 그 뒤에 객체 지향을 지키는 게 맞는 모습인 것 같습니다. 그래서 한 번의 쿼리로 조회할 수 있도록 수정해주었습니다!

jinyoungchoi95 commented on src/main/java/roomescape/common/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,19 @@
+        String title = "RoomEscape Application Swagger";
+        String description = "방탈출 서비스의 API 문서입니다.";
+
+        Info info = new Info().title(title).description(description).version("1.0.0");
```
`1.0.0`의 기준은 무엇인가요? 서버의 배포버전인가요?

theminjunchoi replied:
저 버전은 Swagger UI에서 보여줄 문서의 버전입니다. 
꼭 서버의 배포 버전과 1대1로 매칭되어야할 필요는 없는 걸로 알고 있어요.

jinyoungchoi95 replied:
매칭되어야할 필요는 없지만 매칭되지 않아야하는 이유도 없는 것 같아요.

서버 문서를 보는 입장에서 swagger는 `1.0.0`이고 api는 `version = '0.0.1-SNAPSHOT'`라면 다른 사람이 와서 코드를 봤을때 신뢰할 수 있을까요?

theminjunchoi replied:
> 서버 문서를 보는 입장에서 swagger는 1.0.0이고 api는 version = '0.0.1-SNAPSHOT'라면 다른 사람이 와서 코드를 봤을때 신뢰할 수 있을까요?

말씀하신 경우는 신뢰할 수 없을 것 같아요. 사용자가 문서 버전을 보고 안정 버전인 줄 알고 쓰지만 실제로는 아직 개발 중일 수도 있을 것 같습니다.

그래서 문서 버전이 실제 api 배포 버전을 따라가거나, 굳이 다르게 해주고 싶으면 doc-0.1 과 같은 식으로 단순 문서만의 버전임을 명시해주면 좋을 것 같아요

jinyoungchoi95 commented on src/main/java/roomescape/payment/domain/Payment.java:
```diff
@@ -28,20 +29,26 @@ public class Payment {
 
     @Enumerated(value = EnumType.STRING)
-    private PaymentStatus status;
+    private PaymentStatus paymentStatus;
```
위에서 객체지향에 대한 질문을 남겨두셔서 추가로 코멘트를 남겨보면, 현재 코드는 객체지향인가요? 상태를 들고있는 것이 아닌 해당 클래스를 추상화하고 status에 따른 구현체가 있었어야하지 않을까요?

theminjunchoi replied:
짚어주신 부분도 고민을 했었어요.

사실 상태에 따라 책임이 일부 다르기 때문에 객체를 나눠서 구분해주는 게 객체지향의 관점에서는 올바른 행동이라 생각합니다. 

하지만 지금 구조에선 상태마다 클래스를 분리하면 도메인 객체가 지나치게 세분화돼서 오히려 관리가 더 어려워질 수도 있다고 봤어요.
예를 들어 PendingPayment, SuccessPayment, FailedPayment 이런 식으로 나누면 JPA 매핑이 복잡해지고, 상태 전이 로직도 분산되다 보니 가독성이나 유지보수 측면에서 오히려 손해일 것 같았습니다.

그래서 지금은 상태에 따른 유효성 검사만 간단히 메서드로 때우는 게 더 깔끔하다고 판단했어요.
근데 상태마다 행동이 더 복잡해지고, 정책이 달라지는 등, 부여되는 책임이 많이 다르면 그때는 전략 패턴 등을 적용하면서 책임 분리하는 쪽으로 리팩토링할 생각입니다.

저도 답변을 적으면서 생각이 든건, 결국 “객체지향적으로 완벽하게 분리하는 게 무조건 옳다”보다, 지금 상황에서 복잡도 대비 효용이 크냐 작냐 이게 더 중요하다고 생각이 들었습니다. 그래서 일단은 enum + 상태 전이 메서드 정도에서 마무리했습니다.

jinyoungchoi95 replied:
객체지향이란게 어떤 것이고 프로젝트를 진행하면서 생각하던 객체지향을 못지키는 경우도 발생하게 될건데 그런부분들에서 지금처럼 트레이드오프를 잘 고려해보시면 프로젝트를 진행하시는데 도움이 될 것 같아요.

jinyoungchoi95 commented on src/main/java/roomescape/payment/ui/PaymentController.java:
```diff
@@ -12,13 +13,14 @@
         this.paymentService = paymentService;
     }
 
+    @Operation(summary = "결제 데이터 저장 API")
```
api 문서에는 `Operation`만 있으면 될까요? 좋은 api 문서란 어떤 것일까요? 현재 공유주신 swagger만 읽고도 처음 api를 사용하는 개발자가 불편한점은 없나요?

theminjunchoi replied:
"처음보는 개발자도 아무한테도 안 물어보고 api를 제대로 쓸 수 있게 해주는 문서" 가 제가 생각했을 때 좋은 api 문서라고 생각해요.

지금 코드에서도 `@Operation`에 description을 추가한다든가, `@ApiResponse`를 통해 성공/실패 케이스를 명시하고, 또는 DTO에 `@Schema`로 설명을 추가하고, 쿼리 파라미터가 있는 경우 `@Parameter`로 추가 정보를 명시해 줄 수 있지만 다 추가해줬을 경우 코드 가독성이 너무 저하되고 유지보수 관점에서도 비용이 너무 클 것 같아요. 

그래서 어떤 api가 있는지와 같은 핵심만 전달하고자 했습니다. 세부 내용도 전달해야한다면, Postman이나 수기로 작성할 수 있을 것 같고, 팀 프로젝트 단위라면 notion으로 정리해두는 것도 방법일 것 같아요.

jinyoungchoi95 replied:
받는 입장이 되어서 생각해보시면 좋겠어요.

토스 api를 사용하실때 어떻던가요? 물론 그게 static한 문서이기는하지만 토스 문서에서 각 케이스마다 상황설명을 해주지 않았으면 어떨때 어떤 값이 내려온다를 알 수 있었을까요?

아마 우테코 프로젝트에서까지도 그럴일은 없겠지만 서버도 다른 서버를 호출받아야해서 그 문서를 봐야할때가 있어요. 그럴때 내가 어떤 문서를 받으면 좋을까를 고민해보면서 문서에 어떤 것까지 작성해줘야하는지 정해보시면 좋을 것 같아요.

postman이나 notion으로 더 쓸수는 있지만 그건 관리포인트가 두배로 늘어나게 될 것 같네요. 저는 관리되는 문서는 swagger하나로도 충분하다고 생각해요.

jinyoungchoi95 commented on build.gradle:
```diff
@@ -24,6 +24,8 @@ repositories {
 }
 
 dependencies {
+    implementation group: 'org.springdoc', name: 'springdoc-openapi-starter-webmvc-ui', version: '2.8.8'
```
openapi와 swagger를 사용하신 이유가 무엇인가요?

theminjunchoi replied:
OpenAPI는 REST API를 표준화된 방식으로 명세할 수 있어서, API의 구조와 동작을 문서로 명확하게 표현할 수 있다는 점이 장점이라고 생각했습니다.  

Swagger는 그 OpenAPI 명세를 기반으로 자동으로 문서를 생성하고, 시각화된 UI로 테스트까지 할 수 있게 해주기 때문에 개발자 간 협업에서 큰 이점이 있다고 판단했습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 엠제이 😄

방탈출 결제 / 배포 미션의 리뷰어를 맡은 오리입니다.

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

theminjunchoi committed:
refactor: 결제 재시도 로직 및 관련 테스트 코드 제거

theminjunchoi committed:
refactor: PaymentRepository 메서드 리팩토링 및 MyHistory 조회 로직 개선

theminjunchoi committed:
refactor: 로그 AOP 추가 및 예외 처리 로직 개선

- Controller 및 Service에 대한 AOP 기반 로그 처리 추가
- 전역 예외 핸들러의 로깅 방식 개선 (lombok의 @Slf4j 사용)
- 예외 발생 시 메서드, URI, 예외 타입 등의 상세 정보를 포함한 로그 포맷 적용

jinyoungchoi95 commented on src/main/java/roomescape/common/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class SwaggerConfig {
```
Swagger에는 문서로써 표현할 수 있는 옵션들이 더 다양하기때문에 어떤 옵션들이 있는지도 한번쯤 확인해보시면 좋을 것 같아요. 개인적으로 신뢰가는 개발자는 상대방을 위하는 문서/글을 작성할 줄 아는 개발자라고 생각합니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 엠제이 😄

현재 미션에서 필요한 학습을 모두 하신 것 같아 머지하려하는데요. 남겨둔 코멘트에 궁금하신점이 있을까 하여 Request Changes 드려요.
코멘트는 모두 반영할 필요는 없고 다시 리뷰요청주시면 머지하겠습니다.

감사합니다 🙇‍♂️

theminjunchoi committed:
chore: Swagger API 버전 변경

theminjunchoi commented:
안녕하세요 오리..!
리뷰 감사합니다!

6/10, 23:10 기준으로 더 이상 질문이 없어서 Swagger 버전 변경 및 문서에 추가 정보 띄워주는 것 정도만 수정해주었습니다!

jinyoungchoi95 reviewed (approved):
안녕하세요 엠제이 😄

방탈출 결제 미션기간동안 주고받았던 피드백이 앞으로도 많은 도움이 되길 바랍니다 😄
2레벨 고생하셨어요.

감사합니다 🙇‍♂️
